File: src/main/java/org/manapart/lumberjack/Lumberjack.java
package org.manapart.lumberjack;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.entity.item.ItemEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IWorld;
import net.minecraft.world.server.ServerWorld;
import net.minecraft.world.storage.loot.LootContext;
import net.minecraft.world.storage.loot.LootParameters;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.ToolType;
import net.minecraftforge.event.world.BlockEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

import java.util.ArrayList;
import java.util.List;

// The value here should match an entry in the META-INF/mods.toml file
@Mod("lumberjack")
public class Lumberjack {

    public Lumberjack() {
        MinecraftForge.EVENT_BUS.register(this);
    }

    @SubscribeEvent
    public void onBreak(BlockEvent.BreakEvent event) {
        Block block = event.getState().getBlock();
        BlockPos pos = event.getPos();

        if (shouldFellTrees(event.getPlayer())) {
            if (isLog(block)) {
                System.out.println(event.getPlayer().getName() + " broke " + block + " at " + pos);
                fellLogs(pos, event.getWorld(), event.getPlayer().getHeldItemMainhand());
            }
        }
    }

    public static boolean isLog(Block block) {
        return isType(block, "log");
    }

    public static boolean isLeaves(Block block) {
        return isType(block, "leaves");
    }

    private static boolean isType(Block block, String type) {
        if (block.getRegistryName() != null) {
            return block.getRegistryName().getPath().toLowerCase().contains(type);
        }
        return false;
    }

    private boolean shouldFellTrees(PlayerEntity player) {
        Item item = player.getHeldItemMainhand().getItem();
        return !player.isSneaking() && item.getToolTypes(player.getHeldItemMainhand()).contains(ToolType.AXE);
    }


    private void fellLogs(BlockPos sourcePosition, IWorld world, ItemStack tool) {
        ColumnFinder finder = new ColumnFinder(sourcePosition, world);
        ArrayList<BlockPos> columns = finder.findColumns();
        chopColumns(columns, 1, world, tool);
    }

    private void chopColumns(ArrayList<BlockPos> columns, int y, IWorld world, ItemStack tool) {
        boolean atLeastOneBlockHarvested = false;
        for (BlockPos column : columns) {
            BlockPos pos = column.add(0, y, 0);
            BlockState blockState = world.getBlockState(pos);
            Block block = blockState.getBlock();

            if (isLog(block) || isLeaves(block)) {
                dropBlock(world, pos, blockState, tool);
                atLeastOneBlockHarvested = true;
            }
        }
        if (atLeastOneBlockHarvested) {
            chopColumns(columns, y + 1, world, tool);
        }
    }

    private void dropBlock(IWorld world, BlockPos pos, BlockState state, ItemStack tool) {
        if (world instanceof ServerWorld) {
            LootContext.Builder lootContext = new LootContext.Builder((ServerWorld) world);
            lootContext.withParameter(LootParameters.TOOL, tool);
            lootContext.withParameter(LootParameters.POSITION, pos);
            List<ItemStack> drops = state.getDrops(lootContext);
            world.removeBlock(pos, false);

            if (drops.size() >= 1) {
                dropItems(world, pos, drops);
            }
        }
    }

    private void dropItems(IWorld world, BlockPos pos, List<ItemStack> drops) {
        for (ItemStack i : drops) {
            if (i != null) {
                ItemEntity dropItem = new ItemEntity(world.getWorld(), (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, i);
                world.addEntity(dropItem);
            }
        }
    }


}

File: src/main/kotlin/org/manapart/lumberjack/Lumberjack.kt
package org.manapart.lumberjack

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.event.player.PlayerBlockBreakEvents
import net.minecraft.core.BlockPos
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.AxeItem
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import org.slf4j.LoggerFactory

object Lumberjack : ModInitializer {
    private val log = LoggerFactory.getLogger("lumberjack")

    override fun onInitialize() {
        PlayerBlockBreakEvents.AFTER.register(::onBreak)
    }

    private fun onBreak(level: Level, player: Player, pos: BlockPos, state: BlockState, entity: BlockEntity?) {
        if (shouldFellTrees(player)) {
            if (state.block.isLog()) {
//                log.info(player.name.contents.toString() + " broke " + state.block + " at " + pos)
                fellLogs(pos, TestableWorld(level), player.mainHandItem)
            }
        }
    }

    private fun shouldFellTrees(player: Player): Boolean {
        return !player.isCrouching && player.mainHandItem.item is AxeItem
    }

    fun fellLogs(sourcePosition: BlockPos, world: TestableWorld, tool: ItemStack?) {
//        val finder = ColumnFinder(sourcePosition, world)
//        val columns = finder.findColumns()
//        chopColumns(columns, 1, world, tool)
        val (logs, leaves) = climbTree(world, sourcePosition)
        chopTree(world, logs, leaves, tool)
    }

//    private fun chopColumns(columns: List<BlockPos>, y: Int, world: TestableWorld, tool: ItemStack?) {
//        var atLeastOneBlockHarvested = false
//        for (column in columns) {
//            val pos = column.offset(0, y, 0)
//            if (world.isLog(pos) || world.isLeaves(pos)) {
//                dropBlock(world, pos, tool)
//                atLeastOneBlockHarvested = true
//            }
//        }
//        if (atLeastOneBlockHarvested) {
//            chopColumns(columns, y + 1, world, tool)
//        }
//    }
//
//    private fun dropBlock(world: TestableWorld, pos: BlockPos, tool: ItemStack?) {
//        if (world.level != null && world.level is ServerLevel) {
//            val origin = Vec3.atCenterOf(pos)
//            world.removeBlock(pos, false)
//            if (tool != null) {
//                val lootContext = LootParams.Builder(world.level)
//                    .withParameter(LootContextParams.TOOL, tool)
//                    .withParameter(LootContextParams.ORIGIN, origin)
//
//                val drops = world.level.getBlockState(pos).getDrops(lootContext).filterNotNull()
//                if (drops.isNotEmpty()) {
//                    dropItems(world.level, pos, drops)
//                }
//            }
//        } else {
//            world.removeBlock(pos, false)
//        }
//    }
//
//    private fun dropItems(world: Level, pos: BlockPos, drops: List<ItemStack>) {
//        drops.forEach { drop ->
//            val dropItem = ItemEntity(world, pos.x.toDouble() + 0.5, pos.y.toDouble() + 0.5, pos.z.toDouble() + 0.5, drop)
//            world.addFreshEntity(dropItem)
//        }
//    }

}

fun Block.isLog(): Boolean {
    return isType(this, "log")
}

fun Block.isLeaves(): Boolean {
    return isType(this, "leaves")
}

private fun isType(block: Block, type: String): Boolean {
    return block.descriptionId.lowercase().contains(type)
}
